File: android/core/src/main/kotlin/com/ridelink/core/protocol/Dedup.kt
package com.ridelink.core.protocol

import com.ridelink.core.model.ConnTiebreak
import com.ridelink.core.model.PeerId

/**
 * PROTOCOL §4.2 / ADR-015. Deliberately uses [ConnTiebreak], never [PeerId] — see
 * ADR-015 "Why conn_tiebreak and not peer_id".
 *
 * ADR-015 Amendment A2 / ADR-010 Amendment A2 (this session's correction): connection ownership
 * and leadership are independent by construction. Nothing in this object may be used to infer,
 * or be inferred from, [Leadership]. **No caller may assume `initiator == leader` or
 * `acceptor == leader`.**
 */
object Dedup {
    /** Comparison is over the 32-character lowercase hex string (PROTOCOL §4.2). */
    sealed class Verdict {
        /** The connection initiated by [survivorPeer] survives; the other peer's outbound is closed. */
        data class Survivor(
            val survivorPeer: Side,
        ) : Verdict()

        /** conn_tiebreak values are byte-identical (probability 2^-128): both sides close and retry. */
        object Tie : Verdict()
    }

    enum class Side { A, B }

    data class PeerTiebreak(
        val peerId: PeerId,
        val connTiebreak: ConnTiebreak,
    )

    /**
     * @return which side's *outbound* connection survives. The rule: the surviving connection is
     *   the one initiated by the peer with the **larger** conn_tiebreak.
     */
    fun resolve(
        a: PeerTiebreak,
        b: PeerTiebreak,
    ): Verdict {
        val cmp = a.connTiebreak.value.compareTo(b.connTiebreak.value)
        return when {
            cmp == 0 -> Verdict.Tie
            cmp > 0 -> Verdict.Survivor(Side.A)
            else -> Verdict.Survivor(Side.B)
        }
    }
}

/** ADR-010. Keyed on [PeerId] alone — never on [ConnTiebreak] (see [Dedup]'s doc comment). */
object Leadership {
    /** The lexicographically smaller peer_id leads. */
    fun elect(
        a: PeerId,
        b: PeerId,
    ): Dedup.Side = if (a.value < b.value) Dedup.Side.A else Dedup.Side.B
}

File: ios/Packages/RideLinkCore/Sources/RideLinkCore/Protocol/Dedup.swift
import Foundation

/// PROTOCOL §4.2 / ADR-015. Deliberately uses `ConnTiebreak`, never `PeerId` — see
/// ADR-015 "Why conn_tiebreak and not peer_id".
///
/// ADR-015 Amendment A2 / ADR-010 Amendment A2 (this session's correction): connection ownership
/// and leadership are independent by construction. Nothing in this type may be used to infer, or
/// be inferred from, `Leadership`. **No caller may assume `initiator == leader` or
/// `acceptor == leader`.**
public enum Dedup {
    public enum Side: Sendable, Equatable {
        case a
        case b
    }

    /// Comparison is over the 32-character lowercase hex string (PROTOCOL §4.2).
    public enum Verdict: Sendable, Equatable {
        /// The connection initiated by the associated side survives; the other peer's outbound is closed.
        case survivor(Side)
        /// conn_tiebreak values are byte-identical (probability 2^-128): both sides close and retry.
        case tie
    }

    public struct PeerTiebreak: Sendable {
        public let peerId: PeerId
        public let connTiebreak: ConnTiebreak

        public init(peerId: PeerId, connTiebreak: ConnTiebreak) {
            self.peerId = peerId
            self.connTiebreak = connTiebreak
        }
    }

    /// - Returns: which side's *outbound* connection survives. The rule: the surviving connection
    ///   is the one initiated by the peer with the **larger** conn_tiebreak.
    public static func resolve(_ a: PeerTiebreak, _ b: PeerTiebreak) -> Verdict {
        if a.connTiebreak.value == b.connTiebreak.value {
            return .tie
        }
        return a.connTiebreak.value > b.connTiebreak.value ? .survivor(.a) : .survivor(.b)
    }
}

/// ADR-010. Keyed on `PeerId` alone — never on `ConnTiebreak` (see `Dedup`'s doc comment).
public enum Leadership {
    /// The lexicographically smaller peer_id leads.
    public static func elect(_ a: PeerId, _ b: PeerId) -> Dedup.Side {
        a.value < b.value ? .a : .b
    }
}
